### PR TITLE
[MRG+1] Renderer class as visualization backend abstraction

### PR DESCRIPTION
**Pull request by GuillaumeFavelier**
#### What does this implement/fix?
This goal of this PR is to isolate actual implementation of visualization features as backend file stored in the new `backend` directory so that it's easier to maintain this part of the code and manage potentially new backends.

#### Additional information
I started the code refactoring of the function `plot_alignment()`, it's still a WIP. I'm currently working on the `mayavi.mlab.quiver3d()` function.

### CODE REVIEW

File: mne/__init__.py [hunk 1]
```diff
@@ -18,6 +18,42 @@
...
 
+# Set the default backend
+import os
+import importlib
+default_backend = 'mlab'
+try:
+    if _BACKEND is None:
+        _BACKEND = os.environ.get('MNE_BACKEND', default_backend)
+except NameError:
+    _BACKEND = os.environ.get('MNE_BACKEND', default_backend)
```
Comment (agramfort) [inline mne/__init__.py]: call it MNE_3D_BACKEND

File: mne/__init__.py [hunk 2]
```diff
@@ -18,6 +18,42 @@
...
+
+
+def set_backend(backend_name):
+    """Sets the backend for MNE
+        The backend will be set as specified and operations will used
+        that backend
```
Comment (agramfort) [inline mne/__init__.py]: check pep257

File: mne/__init__.py [hunk 3]
```diff
@@ -18,6 +18,45 @@
 
 __version__ = '0.18.dev0'
 
+# Set the default backend
```
Comment (massich) [inline mne/__init__.py]: I would move all this into some existing utils. ie `.utils` `.viz.utils` and import it.

If does not make sense in any of the existing utils, I would create something new. If you want to keep it separate it while developing, I would put in a new file somewhere else while we figure out where shall it go.

Comment (GuillaumeFavelier) [reply mne/__init__.py]: Thanks for the feedback! I like the idea. I'm gonna work on it.

File: mne/viz/utils.py [hunk 4]
```diff
@@ -43,6 +45,46 @@
...
+
+global _BACKEND
+try:
+    if _BACKEND is None:
+        _BACKEND = os.environ.get('MNE_3D_BACKEND', default_backend)
+except NameError:
+    _BACKEND = os.environ.get('MNE_3D_BACKEND', default_backend)
+
+
+def set_backend(backend_name):
```
Comment (massich) [inline mne/viz/utils.py]: all these function names are a bit confusing to me (as a user). 

I guess that `set_backend`, `get_backend` etc are exposed in viz. Then the user code:

```py
from mne.viz import set_backend

set_backend(my_awesome_backend)
```
with this code I don't really understand which kind of backend. Either you add some info in the function name, or you put them in a directory.

```py
from mne.viz.3D import set_backend
set_backend(my_awesome_backend) # I know I'm talking aobut 3D 'cos I've been the one importing it.
```

```py
from mne.viz import set_3D_backend
set_3D_backend(my_awesome_backend) # I know I'm talking aobut 3D 'cos the name says so.
```

If you ask me I like the first one better.

Comment (GuillaumeFavelier) [reply mne/viz/utils.py]: Good idea, thanks! I thought convenient to have it all in `mne.viz` to be honest so I can implement your second idea and wait for more feeback.

File: mne/viz/mlab_backend.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+def add_arrows(x, y, z, u, v, w, color, scale, resolution, opacity=1.0,
+               scale_mode='none', scalars=None):
+    renderer.mlab.quiver3d(x, y, z, u, v, w, mode='arrow', scale_factor=scale,
+                           color=color, scale_mode=scale_mode,
+                           resolution=resolution, scalars=scalars,
+                           opacity=opacity, figure=renderer.fig)
+
+
+def add_3d_arrows(x, y, z, u, v, w, color, scale, height, resolution,
```
[the remark this answers is not in the document]
Comment (GuillaumeFavelier) [reply mne/viz/mlab_backend.py]: and what would you say about `add_arrows()`, maybe just `add_quiver()` is enough?

File: mne/viz/utils.py (hunk 4 above)
Comment (massich) [reply mne/viz/utils.py]: Don't rush it.

How is the structure of `mne.viz`?

**Comment by agramfort** [comment]
> I would make a mne/viz/backends/ directory and put all your code there. I would have one py file per backend that contained the abstracted functions/methods for it, starting with mne/viz/backends/_pysurfer_mayavi.py or so

I agree with this

**Comment by GuillaumeFavelier** [comment]
Following @larsoner advice, I moved the backend implementation in `mne/viz/backends` directory. 

As of now, changing the backend is done as follows :
```python
from mne.viz.backends.renderer import set_3D_backend
set_3D_backend(my_awesome_backend)
```

File: mne/viz/_3d.py [hunk 6]
```diff
@@ -581,16 +582,13 @@ def plot_alignment(info, trans=None, subject=None, subjects_dir=None,
         It corresponds to Freesurfer environment variable SUBJECTS_DIR.
     surfaces : str | list
         Surfaces to plot. Supported values:
-
```
Comment (agramfort) [inline mne/viz/_3d.py]: @GuillaumeFavelier can you avoid such cosmetic changes? it makes diff longer to read for little gain. thx

File: mne/viz/_3d.py [hunk 7]
```diff
@@ -626,50 +624,35 @@ def plot_alignment(info, trans=None, subject=None, subjects_dir=None,
...
         .. versionadded:: 0.16
     verbose : bool, str, int, or None
         If not None, override default verbose level (see :func:`mne.verbose`
         and :ref:`Logging documentation <tut_logging>` for more).
-
-    Returns
-    -------
-    fig : instance of mlab.Figure
-        The mayavi figure.
```
Comment (agramfort) [inline mne/viz/_3d.py]: if you don't return anything anymore it's an API break. Careful with this.

File: mne/viz/_3d.py [hunk 8]
```diff
@@ -1061,12 +1044,8 @@ def plot_alignment(info, trans=None, subject=None, subjects_dir=None,
                              for pick in seeg_picks])
 
     # initialize figure
-    if fig is None:
-        fig = _mlab_figure(bgcolor=(0.5, 0.5, 0.5), size=(800, 800))
-    if interaction == 'terrain' and fig.scene is not None:
-        fig.scene.interactor.interactor_style = \
-            tvtk.InteractorStyleTerrain()
-    _toggle_mlab_render(fig, False)
+    renderer.init((800, 800), (0.5, 0.5, 0.5))
```
Comment (agramfort) [inline mne/viz/_3d.py]: please keep keywords names explicit.

File: mne/viz/_3d.py [hunk 9]
```diff
@@ -1075,19 +1054,17 @@ def plot_alignment(info, trans=None, subject=None, subjects_dir=None,
...
     for key, surf in surfs.items():
-        # Make a solid surface
-        mesh = _create_mesh_surf(surf, fig)
         with warnings.catch_warnings(record=True):  # traits
-            surface = mlab.pipeline.surface(
-                mesh, color=colors[key], opacity=alphas[key], figure=fig)
-        if key != 'helmet':
-            surface.actor.property.backface_culling = True
+            if key != 'helmet':
+                renderer.add_surface(surf, colors[key], alphas[key], True)
```
Comment (agramfort) [inline mne/viz/_3d.py]: try to keep explicit parameter names. For example here True it's not clear what it means. If you name it it becomes clear.

File: mne/viz/_3d.py [hunk 10]
```diff
@@ -1132,44 +1110,33 @@ def plot_alignment(info, trans=None, subject=None, subjects_dir=None,
...
             with warnings.catch_warnings(record=True):  # traits
-                points = mlab.points3d(data[:, 0], data[:, 1], data[:, 2],
-                                       color=color, scale_factor=scale,
-                                       opacity=alpha, figure=fig)
-                points.actor.property.backface_culling = True
+                renderer.add_spheres(data, color, scale, alpha, True)
     if len(eegp_loc) > 0:
         with warnings.catch_warnings(record=True):  # traits
-            quiv = mlab.quiver3d(
+            renderer.add_3d_arrows(
```
Comment (agramfort) [inline mne/viz/_3d.py]: if 3d arrows is a vispy naming convention why not but if not I would suggest to keep names that are standard for vtk/mayavi. So quiver3d

File: mne/viz/backends/renderer.py [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+    """Set the backend for MNE.
+
+    The backend will be set as specified and operations will use
+    that backend
+
+    Parameters
+    ----------
+    backend_name : {'mlab'}, default is 'mlab'
+    """
+    global MNE_3D_BACKEND
```
Comment (agramfort) [inline mne/viz/backends/renderer.py]: tensorly also uses a global?

thinking out loud maybe a context manage at some point would be nice to have:

```
with mne.3d_backend('mayavi'):
    ....
```

File: mne/viz/_3d.py (hunk 6 above)
Comment (GuillaumeFavelier) [reply mne/viz/_3d.py]: Yes of course. my bad, I'll be more careful.

File: mne/viz/_3d.py (hunk 7 above)
Comment (GuillaumeFavelier) [reply mne/viz/_3d.py]: what do you suggest to return then? `mlab.Figure` is backend specific.

File: mne/viz/_3d.py (hunk 9 above)
Comment (GuillaumeFavelier) [reply mne/viz/_3d.py]: I'm working on it.

File: mne/viz/backends/renderer.py (hunk 11 above)
Comment (GuillaumeFavelier) [reply mne/viz/backends/renderer.py]: tensorly uses global in `set_backend()` and `get_backend()` as well.

File: mne/viz/_3d.py (hunk 8 above)
Comment (GuillaumeFavelier) [reply mne/viz/_3d.py]: I tried to find a generic name for a function making basic pre-configuration of the visualization.

File: mne/viz/_3d.py (hunk 10 above)
Comment (GuillaumeFavelier) [reply mne/viz/_3d.py]: ok, I'll modify it accordingly

**Comment by massich** [comment]
If you reset you oblige me to collect all my data and supply it at once. Rather than add this and add that and add.. in the scene.

**Comment by agramfort** [comment]
keep the pace @GuillaumeFavelier ! no strong feeling about the add_ or not. For me it's a detail. Let's make it work first and then we make it nice.

**Comment by massich** [comment]
Yes. Just let it as an open discussion and we can remove the commit or do a
find replace.

The idea is just to keep bouncing ideas and sharing points of view. By the
end we'll call plot.

File: mne/viz/_3d.py (hunk 7 above)
Comment (GuillaumeFavelier) [reply mne/viz/_3d.py]: What do you think of returning the instance of the Renderer itself? That would be really convenient considering the current architecture.

**Comment by agramfort** [comment]
sounds like a good idea

File: mne/viz/_3d.py [hunk 12]
```diff
@@ -560,9 +552,9 @@ def _plot_mri_contours(mri_fname, surf_fnames, orientation='coronal',
 @verbose
 def plot_alignment(info, trans=None, subject=None, subjects_dir=None,
                    surfaces='head', coord_frame='head',
-                   meg=None, eeg='original',
+                   meg=None, eeg='original', ren=None,
                    dig=False, ecog=True, src=None, mri_fiducials=False,
-                   bem=None, seeg=True, show_axes=False, fig=None,
```
Comment (agramfort) [inline mne/viz/_3d.py]: @GuillaumeFavelier be careful with API breakage. We have a strong policy of deprecation warnings for one release cycle. Any code that works with 0.17 should work with 0.18 out of the box but with warnings in case of deprecation.

File: mne/viz/backends/_pysurfer_mayavi.py [hunk 13]
```diff
@@ -65,10 +65,34 @@ def set_interactive():
...
+def contour(surface, scalars, contours, line_width=1.0, opacity=1.0,
+            vmin=None, vmax=None, colormap=None):
+    mesh = _create_mesh_surf(surface, renderer.fig, scalars=scalars)
+    cont = renderer.mlab.pipeline.contour_surface(
+        mesh, contours=contours, line_width=1.0, vmin=vmin, vmax=vmax,
+        opacity=opacity, figure=renderer.fig)
+    cont.module_manager.scalar_lut_manager.lut.table = colormap
+
+
+def surface(surface, color=(0.7, 0.7, 0.7), opacity=1.0,
```
Comment (agramfort) [inline mne/viz/backends/_pysurfer_mayavi.py]: what is the conceptual difference between surface and mesh?

Comment (GuillaumeFavelier) [reply mne/viz/backends/_pysurfer_mayavi.py]: I would say surface has a more generic definition, you can define a surface by an equation for example while a mesh will be described by its combinatorial elements : vertice, edge and triangle.

**Comment by agramfort** [comment]
We need both in public api?

File: mne/viz/_3d.py (hunk 12 above)
Comment (GuillaumeFavelier) [reply mne/viz/_3d.py]: I can rollback `_3d.py` and think about a slower strategy of integration.

**Comment by GuillaumeFavelier** [comment]
I don't think so.

**Comment by GuillaumeFavelier** [comment]
I got a timeout in `time_scale_mri()` in `mne/tests/test_coreg.py`. That's why Travis is not happy.

**Comment by agramfort** [comment]
I restarted the CI

**Comment by GuillaumeFavelier** [comment]
Many thanks.

**Comment by GuillaumeFavelier** [comment]
According to Travis, some tests failed because of timeouts: `test_compute_proj_parallel`, `test_megsim` and `test_downloads`.

**Comment by GuillaumeFavelier** [comment]
Even if `importlib` reload the backend, the instances will keep the previous `Renderer` definition, that's why the backend need to be imported on-demand and thus updated.

**Comment by GuillaumeFavelier** [comment]
what do you think? @massich @larsoner @agramfort

File: mne/viz/_3d.py [hunk 14]
```diff
@@ -636,6 +614,15 @@ def plot_alignment(info, trans=None, subject=None, subjects_dir=None,
         Mayavi Scene in which to plot the alignment.
         If ``None``, creates a new 600x600 pixel figure with black background.
 
+        .. versionadded:: 0.18
```
Comment (agramfort) [inline mne/viz/_3d.py]: Version added should be below parameter definition

File: mne/viz/backends/_pysurfer_mayavi.py [hunk 15]
```diff
@@ -0,0 +1,419 @@
...
+Actual implementation of Renderer and Projection classes.
+"""
+
+import warnings
+import numpy as np
+from ...surface import _normalize_vectors
+from ...utils import _import_mlab, _validate_type
+
+
+class Projection:
```
Comment (agramfort) [inline mne/viz/backends/_pysurfer_mayavi.py]: I have the habit it always inherit from object

File: mne/viz/backends/renderer.py [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+    MNE_3D_BACKEND = backend_name
+    from . import renderer
+    importlib.reload(renderer)
+
+
+def get_3D_backend():
```
Comment (agramfort) [inline mne/viz/backends/renderer.py]: I would write 3d and not 3D to respect snake case

File: mne/viz/_3d.py (hunk 14 above)
Comment (GuillaumeFavelier) [reply mne/viz/_3d.py]: whoops, my bad.

File: mne/viz/backends/renderer.py [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+    MNE_3D_BACKEND = os.environ.get('MNE_3D_BACKEND', default_3d_backend)
+
+if MNE_3D_BACKEND == 'mayavi':
+    from ._pysurfer_mayavi import Renderer, Projection  # noqa: F401
+else:
+    import warnings
+    warnings.warn('MNE_3D_BACKEND should be "mayavi" : '
+                  '{} was given.'.format(MNE_3D_BACKEND))
+
+sys.stderr.write('Using {} backend.\n'.format(MNE_3D_BACKEND))
```
Comment (agramfort) [inline mne/viz/backends/renderer.py]: I would use the mne logging system here. It’s an info message

File: mne/viz/backends/renderer.py [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+
+try:
+    if MNE_3D_BACKEND is None:
+        MNE_3D_BACKEND = os.environ.get('MNE_3D_BACKEND',
+                                        default_3d_backend)
+except NameError:
+    MNE_3D_BACKEND = os.environ.get('MNE_3D_BACKEND', default_3d_backend)
+
+if MNE_3D_BACKEND == 'mayavi':
+    from ._pysurfer_mayavi import Renderer, Projection  # noqa: F401
```
Comment (agramfort) [inline mne/viz/backends/renderer.py]: Can we make the projection class private? Call it _Projection?

File: mne/viz/backends/renderer.py [hunk 19]
```diff
@@ -0,0 +1,51 @@
+import sys
```
Comment (agramfort) [inline mne/viz/backends/renderer.py]: Add licence and author on top of each file

File: mne/viz/backends/_pysurfer_mayavi.py [hunk 20]
```diff
@@ -0,0 +1,419 @@
+"""
+Core visualization operations based on Mayavi.
+
+Actual implementation of Renderer and Projection classes.
+"""
```
Comment (agramfort) [inline mne/viz/backends/_pysurfer_mayavi.py]: Add licence and author

**Comment by agramfort** [comment]
That’s it for me

@larsoner can you do a pass?

File: mne/viz/_3d.py [hunk 21]
```diff
@@ -637,6 +616,15 @@ def plot_alignment(info, trans=None, subject=None, subjects_dir=None,
...
         .. versionadded:: 0.16
+    renderer : backends.Renderer | None
+        Rendering Scene in which to plot the alignment.
+        If ``None``, creates a new 600x600 pixel figure with black background.
+
+        .. versionadded:: 0.18
+    return_mayavi_figure : bool
+        If True (default), return mayavi figure.
+
+        .. versionadded:: 0.18
```
Comment (larsoner) [inline mne/viz/_3d.py]: I don't understand why there is this new kwarg. Why not always return it?

Comment (larsoner) [reply mne/viz/_3d.py]: (there are a couple of places this was added, so if you revert, please revert in all)

Comment (GuillaumeFavelier) [reply mne/viz/_3d.py]: I introduced it for deprecation reason at first but you're right, it's unnecessary, I'll remove it all.

**Comment by GuillaumeFavelier** [comment]
I'm trying to depreciate this backend-specific variable `fig` of type `mayavi.figure` in `mne.viz._3d.py`. This variable is used as parameter or is returned by mutliple functions of the module. It is a mandatory argument of the function `mne.viz._3d.snapshot_brain_montage()` for example.

So I chose to add a parameter `return_mayavi_figure` always `True` for now for compatibilty reasons to continue returning this `fig` but the goal of this PR is to isolate the backend implementation and be backend-agnostic: I suggest for the future to take a `mne.viz.backends.Renderer` as parameter and return it as a `plot_` result, what do you think @agramfort @larsoner?

**Comment by agramfort** [comment]
as discussed snapshot_brain_montage should accept as input any object
that gets returned by a mne.viz._3d.plot_xxx function. To make the
snapshot_brain_montage backend agnostic it means this object should
know how to do do what is necessary for a snapshot. In other words I
think the we should aim to return a renderer objection in all
mne.viz._3d.plot_ functions and if you have mayavi backend you can get
the fig with renderer._fig if you know what you are doing.

**Comment by larsoner** [comment]
> as discussed snapshot_brain_montage should accept as input any object that gets returned by a mne.viz._3d.plot_xxx function.

This would always be a `mne.viz.Renderer` class, then? Just different subclasses, like `MayaviRenderer` or `VisPyRenderer`?

I'm not sure what the advantage to this is over always returning (and passing in) the root object (figure-like object) for each rendering method object to our `plot_` functions. I guess you'd get to do `my_rend.plot(...)` or something maybe, but I'm not sure that's worth introducing a new class and breaking backward compatibility with our output and input types.

Basically I propose that we return the figure/Brain-like object, and take a figure/Brain-like object from each backend, rather than adding a `mne.viz.XYZRenderer` class whose only job it seems to be is to make our abstracting easier. (This can still be done/used internally, I'm just not sure what the advantage is to making this a user-facing change.)

**Comment by larsoner** [comment]
Sorry to come late to the conversation here, it's possible I'm missing some essential cool feature about the class-based scheme...

**Comment by agramfort** [comment]
how do you make it work with the snapshot_brain_montage function? do
you branch between backend types in snapshot_brain_montage ?
shall we talk about it on hangout with @GuillaumeFavellier?

**Comment by larsoner** [comment]
Yes you'd branch within each function. I'm sure we can find some clever way to keep it one line, a dict-style-lookup based on the class name perhaps.

**Comment by GuillaumeFavelier** [comment]
I compared the images generated by `[circle full]` on both `master` and this PR and I dont see any difference now. What do you think @agramfort @larsoner @massich ?

File: mne/viz/backends/renderer.py [hunk 22]
```diff
@@ -0,0 +1,60 @@
...
+    """Return the backend currently used.
+
+    Returns
+    -------
+    backend_used : str
+        the backend currently in use
+    """
+    global MNE_3D_BACKEND
+    backend_used = MNE_3D_BACKEND
+    return backend_used
```
Comment (larsoner) [inline mne/viz/backends/renderer.py]: Why not just `return MNE_3D_BACKEND`? strings (and None) are immutable so there should be no risk involved

File: mne/viz/backends/renderer.py (hunk 11 above)
Comment (larsoner) [reply mne/viz/backends/renderer.py]: Can you add a smoke test that you can do `set_3d_backend('mayavi')` followed by `set_3d_backend('mayavi')` without it breaking things (do some simple plot afterward)?

And I think that `set_3d_backend('foo')` should be a `ValueError` (rather than `warn`), and `set_3d_backend(1)` should be a TypeError (checked with `mne.utils._validate_type`) and that these degenerate conditions should be checked with `with pytest.raises(..., match='...'):` in a new `mne/viz/backends/tests/test_renderer.py` file.

File: mne/viz/backends/renderer.py [hunk 23]
```diff
@@ -0,0 +1,60 @@
...
+import importlib
+from ..utils import logger
+
+default_3d_backend = 'mayavi'
+
+
+try:
+    if MNE_3D_BACKEND is None:
+        MNE_3D_BACKEND = os.environ.get('MNE_3D_BACKEND',
+                                        default_3d_backend)
```
Comment (larsoner) [inline mne/viz/backends/renderer.py]: Use `get_config('MNE_3D_BACKEND', default_3d_backend)` in these two calls -- it polls `environ` followed by the MNE config file.

File: mne/viz/_3d.py [hunk 24]
```diff
@@ -327,9 +327,11 @@ def plot_evoked_field(evoked, surf_maps, time=None, time_label='t = %0.0f ms',
 
     Returns
     -------
-    fig : instance of mayavi.mlab.Figure
-        The mayavi figure.
+    scene : instance of renderer.scene()
+        The scene.
```
Comment (agramfort) [inline mne/viz/_3d.py]: to revert

File: mne/viz/backends/_pysurfer_mayavi.py [hunk 25]
```diff
@@ -0,0 +1,433 @@
...
+Core visualization operations based on Mayavi.
+
+Actual implementation of Renderer and Projection classes.
```
Comment (agramfort) [inline mne/viz/backends/_pysurfer_mayavi.py]: _Renderer and _Projection

File: mne/viz/backends/_pysurfer_mayavi.py [hunk 26]
```diff
@@ -0,0 +1,433 @@
...
+
+    def quiver3d(self, x, y, z, u, v, w, color, scale, mode, resolution=8,
+                 glyph_height=None, glyph_center=None, glyph_resolution=None,
+                 opacity=1.0, scale_mode='none', scalars=None,
+                 backface_culling=False):
+        """Add quiver3d in the scene.
+
+        Parameters
+        ----------
+        x: array
```
Comment (agramfort) [inline mne/viz/backends/_pysurfer_mayavi.py]: it's a good practive to document the shape eg

`x : array, shape (n_quivers,)`

File: mne/viz/backends/_pysurfer_mayavi.py [hunk 27]
```diff
@@ -0,0 +1,433 @@
...
+            quiv.glyph.glyph_source.glyph_source.center = glyph_center
+            quiv.glyph.glyph_source.glyph_source.resolution = glyph_resolution
+            quiv.actor.property.backface_culling = backface_culling
+
+    def text(self, x, y, text, width):
+        """Add test in the scene.
+
+        Parameters
+        ----------
+        x: array
```
Comment (agramfort) [inline mne/viz/backends/_pysurfer_mayavi.py]: array or float?

Comment (GuillaumeFavelier) [reply mne/viz/backends/_pysurfer_mayavi.py]: float indeed, thank you!

File: mne/viz/_3d.py [hunk 28]
```diff
@@ -648,8 +619,8 @@ def plot_alignment(info, trans=None, subject=None, subjects_dir=None,
 
     Returns
     -------
-    fig : instance of mayavi.mlab.Figure
-        The mayavi figure.
+    scene : instance of renderer.scene()
```
[the remark this answers is not in the document]
Comment (GuillaumeFavelier) [reply mne/viz/_3d.py]: ok, working on it.

**Comment by agramfort** [comment]
there is still hope to merge this before tomorrow morning :)

>

File: mne/viz/backends/_pysurfer_mayavi.py [hunk 29]
```diff
@@ -0,0 +1,433 @@
...
+            The Y component of the position of the quiver.
+        z: array, shape (n_quivers,)
+            The Z component of the position of the quiver.
+        u: array, shape (n_quivers,)
+            The last X component of the quiver.
+        v: array, shape (n_quivers,)
+            The last Y component of the quiver.
+        w: array, shape (n_quivers,)
+            The last Z component of the quiver.
+        color: (red, green, blue)
```
Comment (agramfort) [inline mne/viz/backends/_pysurfer_mayavi.py]: I would write:

```
color: tuple
    The color of the quiver: (red, green, blue).
```

File: mne/viz/backends/_pysurfer_mayavi.py [hunk 30]
```diff
@@ -0,0 +1,433 @@
...
+            The color of the quiver.
+        scale: float
+            The scale of the quiver.
+        mode: 'arrow' or 'cylinder'
+            The type of the quiver.
+        resolution: float
+            The resolution of the arrow.
+        glyph_height: float
+            The height of the glyph used with the quiver.
+        glyph_center:
```
Comment (agramfort) [inline mne/viz/backends/_pysurfer_mayavi.py]: missing

File: mne/viz/backends/_pysurfer_mayavi.py [hunk 31]
```diff
@@ -0,0 +1,433 @@
...
+
+        Parameters
+        ----------
+        azimuth: float
+            The azimuthal angle of the camera.
+        elevation: float
+            The zenith angle of the camera.
+        distance: float
+            The distance to the focal point.
+        focalpoint: (x, y, z)
```
Comment (agramfort) [inline mne/viz/backends/_pysurfer_mayavi.py]: same as above. Say it's a tuple

File: mne/viz/backends/_pysurfer_mayavi.py [hunk 32]
```diff
@@ -0,0 +1,433 @@
...
+        """
+        self.mlab.view(azimuth, elevation, distance,
+                       focalpoint=focalpoint, figure=self.fig)
+
+    def screenshot(self):
+        """Take a screenshot of the scene."""
+        return self.mlab.screenshot(self.fig)
+
+    def project(self, xyz, ch_names):
+        """Convert 3d points to a 2d perspective."""
```
Comment (agramfort) [inline mne/viz/backends/_pysurfer_mayavi.py]: docstring missing

**Comment by agramfort** [comment]
besides LGTM

@GuillaumeFavelier you'll need to rebase on master once the CIs are fixed. Hopefully really soon.

**Comment by larsoner** [comment]
@GuillaumeFavelier you can rebase now that #5915 is in and things should work

**Comment by larsoner** [comment]
FYI @GuillaumeFavelier you will also need to use the two updated hash values from this PR:

https://github.com/mne-tools/mne-python/pull/5904/files#diff-00583e36a75a694e071f9fc054a71c08

I updated the datasets on `osf.io` a couple of days ago. (This is also why `master` CircleCI build currently fails.)

**Comment by larsoner** [comment]
This pull request **introduces 1 alert** when merging 00d978e48985176ff6997d97a8d27e8fb390779c into 63a8ce271eac9b2d488c353e04a3117aa1c5e772 - [view on LGTM.com](https://lgtm.com/projects/g/mne-tools/mne-python/rev/pr-774e76b22c7369890ae3ec5193f76d8427a10dd7)

**new alerts:**

* 1 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by larsoner** [comment]
This pull request **introduces 1 alert** when merging d7bb6bf7288efee0175f7961e9076859e8a08314 into 63a8ce271eac9b2d488c353e04a3117aa1c5e772 - [view on LGTM.com](https://lgtm.com/projects/g/mne-tools/mne-python/rev/pr-d8accf4b0ba584c8a9e4bd309ab98d38145cf4bd)

**new alerts:**

* 1 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by larsoner** [comment]
This pull request **introduces 1 alert** when merging 4658f05f8fff19c2e95139f3881536238af4bebf into 6fdfb41246a1ca1a5ae2c381c5f7d97137a63265 - [view on LGTM.com](https://lgtm.com/projects/g/mne-tools/mne-python/rev/pr-6f484e1092ec7b563ef97846142a9c5cbfb4dead)

**new alerts:**

* 1 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by agramfort** [comment]
Thanks @GuillaumeFavelier !
